### DISCUSSION TITLE
Performance (MP) issues v6 vs. v7 (macOS/MinGW)

### DISCUSSION THREAD

**Discussion by rodlie** [ImageMagick/ImageMagick#3796] [Development]
Hi,

Not sure where to put this at the moment, so discussions it is.

I use Magick++ in several applications and it's great, but when moving from v6 to v7 I noticed major performance (MP) issues on macOS (clang9) and on MinGW (gcc7). Linux and MSVC are unaffected.

Example (on macOS 10.14 using Clang 9.0):
```
convert logo: -scale 4000x4000 tmp.bmp

======================================================
Version: ImageMagick 6.9.12-14 Q16 x86_64 2021-05-31 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(3.1) 
Delegates (built-in): bzlib jng jp2 jpeg lcms lzma png tiff webp zlib
======================================================


real	0m0.805s
user	0m1.957s
sys	0m0.337s

======================================================
Version: ImageMagick 7.0.11-14 Q16 x86_64 2021-05-31 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(3.1) 
Delegates (built-in): bzlib jng jp2 jpeg lcms lzma png tiff webp zip zlib
======================================================


real	0m2.398s
user	0m3.919s
sys	0m0.241s


```

I also tested the first v7 release, and the issue is the same.

Anyone else has this problem?

**Comment by urban-warrior** [comment]
Its a precision verses performance issue. ImageMagick version 7, by default, uses floating point precision.  Build ImageMagick without HDRI support and it will revert to unsigned pixels and restore the performance you expect.

**Comment by rodlie** [comment]
So, v6 HDRI and v7 HDRI is not the same? And why don't I get any performance penalty using v7 on Linux or on Windows using MSVC?

**Reply by urban-warrior** [reply]
The default build of ImageMagick version 6 is non-HDRI.  We now notice your version 6 is HDRI-enabled.  The scaling algorithm is essentially the same between the two versions, except version 6 accumulates each channel red, green, and blue in one pass, version 7 iterates over the channels (channel 0 => 2) individually.  Since there are similar times under Linux and Windows, that suggests its might be a compiler issue.  We assume you built ImageMagick with optimization enabled, perhaps -O3?

**Reply by rodlie** [reply]
Thanks for you feedback.

I will review my builds. Currently the problem is on macOS using Clang from macports (with omp) and Windows using GCC from MinGW/MSYS2. Linux with GCC is ok and Windows with MSVC is ok. And of course v6 is ok on all setups (built identical to v7).
